src/context/MapContext.tsx

```diff
@@ -0,0 +1,126 @@
+  createContext,
+  useRef,
+  useState,
+} from 'react';
+
+import { Extent } from 'ol/extent';
+import BaseLayer from 'ol/layer/Base';
+import Layer from 'ol/layer/Layer';
+import Map from 'ol/Map';
+import { useDebounce } from 'react-use';
+
+import { useFilters } from '@/hooks/useFilters';
+import { getStacCollections } from '@/services/stac';
+import { Collection } from '@/typings/stac';
+
+export type MapContextType = {
+  mapConfig: MapConfig;
+  setMapConfig: Dispatch<SetStateAction<MapConfig>>;
+  mapRef: MutableRefObject<HTMLDivElement | null>;
+  map: Map | undefined;
+  setMap: Dispatch<SetStateAction<Map | undefined>>;
+  getLayers: () => BaseLayer[] | undefined;
+  addLayer: (layer: Layer) => void;
+  removeLayer: (name: string) => void;
+  getLayerByName: (name: string) => BaseLayer | undefined;
+  aoi: Extent | null;
+  setAoi: Dispatch<SetStateAction<Extent | null>>;
+  selectedLayer: Layer | null;
+  setSelectedLayer: Dispatch<SetStateAction<Layer | null>>;
+  collections: Collection[];
+  setCollections: Dispatch<SetStateAction<Collection[] | null>>;
+};
+
+type MapProviderProps = {
+  initialState?: Partial<MapContextType>;
+  children: ReactNode;
+};
+
+export const MapContext = createContext<MapContextType | null>(null);
+MapContext.displayName = 'MapContext';
+
+const DEFAULT_LON_LAT = [-0.09, 51.505];
+const DEFAULT_ZOOM = 7;
+
+export const MapProvider = ({ initialState = {}, children }: MapProviderProps) => {
+  const mapRef = useRef(null);
+
+  const [mapConfig, setMapConfig] = useState<MapConfig>({
+    center: DEFAULT_LON_LAT,
+    zoom: DEFAULT_ZOOM,
+  });
+  const [map, setMap] = useState<Map | undefined>();
+  const [aoi, setAoi] = useState<Extent | null>(null);
+  const [collections, setCollections] = useState<Collection[] | null>(null);
+  const [selectedLayer, setSelectedLayer] = useState<Layer | null>(null);
+
+  const {
+    state: { activeFilters },
+  } = useFilters();
+
+  useDebounce(
+    () => {
+      const fetchData = async () => {
+        try {
+          const collections = await getStacCollections(activeFilters.textQuery);
+          setCollections(collections);
+        } catch (error) {
+          console.error('Error fetching collections', error);
+        }
+      };
+      fetchData();
+    },
+    250,
+    [activeFilters],
+  );
+
+  const getLayers = () => map?.getLayers().getArray();
+
+  const getLayerByName = (name: string) => getLayers()?.find((layer) => layer.get('name') === name);
+
+  const addLayer = (layer: Layer) => {
+    // Ensure the layer is added below the `AOI` drawing layer.
+    layer.setZIndex(5);
+
+    // Add layer to map.
+    map?.addLayer(layer);
+    setSelectedLayer(layer);
+  };
+
+  const removeLayer = (name: string) => {
+    const foundLayer = getLayerByName(name);
+
+    map?.removeLayer(foundLayer as BaseLayer);
+    setSelectedLayer(null);
+  };
+
+  return (
+    <MapContext.Provider
+      value={{
+        mapRef,
+        map,
+        setMap,
```

Comment:
So are we no longer splitting the states into `state` and `actions`?

Reply:
I personally think that complicates things, this is a Context object I've used on several projects, so it is tried and tested.

Most of the contexts in this app are pretty small and in my opinion, the reducer pattern is overkill, whereas a simple useState would suffice.

Reply:
Okay, I don't mind either way. They're definitely small at the moment, unsure on the future size of the project. Regardless, it doesn't make any difference other than consistency.

Reply:
So the docs state `useState` is good for simple isolated data and `useReducer` is good for more interconnected data. What it means by interconnected data is say you have a shopping basket, you would have state something like:

```
const basketState = {
  items,
  totalCost,
  tax,
};

In your reducer when an item is added/amended/removed, you would need to re-calculate the `totalCost`, so state is interconnected.

Having looked at all or context objects, that is not the case, we have a number of bits of state, but they are all independent of each other, not one bit of state relies on another.

One of the big advantages of `useReducer` is that you can just pass the `state` and `dispatch` objects as the Provider value. whereas taking our basket above as an example, using `useState`, we would have to pass bit of state and it's associated setting function, meaning more passed to the Provider. I don't personally see that as a terrible thing, but some do.

However, what we have is a bit of both, we pass the state as recommended by the `useReducer`, but then we wrap the `dispatch` in different setter functions (to hide the complexity of the types and payload), so that is more like the `useState` pattern, so really it is more of a hybrid than actually adhering to either pattern.

The level of state interaction we have at the moment I don't see the need for the complexities of the `useReducer` pattern. If it did grow, I'd probably vote to introduce redux-toolkit, which vastly reduces the boiler-plate logic involved. Also redux has some smart hashing to improve component re-renders, which is an inherent problem with using `Context` manually.

I am happy to discuss this more though.